File: Tide/Tide.Js/src/Cipher.js
// Tide Protocol - Infrastructure for the Personal Data economy
// Copyright (C) 2019 Tide Foundation Ltd
//
// This program is free software and is subject to the terms of
// the Tide Community Open Source License as published by the
// Tide Foundation Limited. You may modify it and redistribute
// it in accordance with and subject to the terms of that License.
// This program is distributed WITHOUT WARRANTY of any kind,
// including without any implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
// See the Tide Community Open Source License for more details.
// You should have received a copy of the Tide Community Open
// Source License along with this program.
// If not, see https://tide.org/licenses_tcosl-1-0-en

import { randomBytes } from "crypto";
import {
  Hash,
  C25519Key,
  AesSherableKey,
  C25519Cipher,
  BnInput,
  Utils,
} from "cryptide";
import Num64 from "./Num64";

export default class Cipher {
  /**
   * @param {string|Uint8Array} data
   * @param {Num64} tag
   * @param {C25519Key} key
   */
  static encrypt(data, tag, key) {
    const buffer = typeof data === 'string' ? Buffer.from(data, "utf-8") : Buffer.from(data);
    
    /**@type Uint8Array */
    let toAsymmetricEncrypt = C25519Cipher.pad(buffer);
    let bufferSymmetric = Buffer.alloc(0);
    
    if (buffer.length > 32) {
      const secret = new AesSherableKey();
      bufferSymmetric = secret.encrypt(buffer);
      toAsymmetricEncrypt = secret.toArray();
    }
    
    const bufferAsymmetric = key.encrypt(toAsymmetricEncrypt).toArray();
    const tagBuffer = tag.toArray();
    const signature = Buffer.from(Utils.padLeft(key.sign(Buffer.concat([bufferAsymmetric, tagBuffer])), 32 * 3));
    
    const size =
      bufferAsymmetric.length +
      tagBuffer.length +
      signature.length +
      bufferSymmetric.length;
    const dimension = dimensionBuffer(size);

    const all = Buffer.alloc(1 + dimension.length + size);

    all[0] = 1; // version #

    dimension.copy(all, 1);
    let step = dimension.length + 1;

    bufferAsymmetric.copy(all, step);
    step += bufferAsymmetric.length;

    tagBuffer.copy(all, step);
    step += tagBuffer.length;

    signature.copy(all, step);
    step += signature.length;

    bufferSymmetric.copy(all, step);
    return all;
  }

  /**
   * @param {Uint8Array} data
   * @param {C25519Key} key
   */
  static decrypt(data, key) {
    let size = data[1] & 127;
    let sizeLength = 0;
    const hasFieldSize = (data[1] & 128) !== 0;
    if (hasFieldSize) {
      sizeLength = size;
      size = BnInput.getBig(data.slice(2, 2 + sizeLength)).valueOf();
    }
    let step = 2 + sizeLength;

    const asymmetricCipher = data.slice(step, step + 32 * 3);
    var asymmetricPlain = key.decrypt(asymmetricCipher);
    step += asymmetricSize();

    if (size === 200)
      return C25519Cipher.unpad(asymmetricPlain);

    const symmetricKey = AesSherableKey.from(asymmetricPlain);
    return symmetricKey.decrypt(data.slice(step));;
  }

  /** @param {Uint8Array} data */
  static asymmetric(data) {
    const step = headEnd(data);
    return data.slice(step, step + asymmetricSize());
  }

  /** @param {Uint8Array} data */
  static symmetric(data) {
    let step = headEnd(data);
    step += asymmetricSize()
    
    return data.slice(step);
  }

  /** @param {Uint8Array} data */
  static cipherFromAsymmetric(data) {
    return C25519Cipher.from(data.slice(0, 32 * 3));
  }
}

/** @param {Uint8Array} data */
function headEnd(data) {
  const sizeLength = (data[1] & 128) !== 0 ? data[1] & 127 : 0;
  return 2 + sizeLength;
}

/**
 * @param {number} size
 * @returns {Uint8Array}
 */
function dimensionBuffer(size) {
  const buffer = BnInput.getArray(BnInput.getBig(size));
  return buffer.length === 1 && buffer[0] < 128
    ? buffer
    : Buffer.concat([Buffer.from([(1 << 7) | buffer.length]), buffer]);
}

/*cipher + signature + tag*/
function asymmetricSize() {
  return 32 * 3 * 2 + 8;
}

File: Tide/Tide.VendorSdk/Classes/Cipher.cs
using System;
using System.Linq;
using System.Numerics;
using System.Text;
using Tide.Encryption.AesMAC;
using Tide.Encryption.Ecc;
using Tide.Encryption.Ed;
using Tide.Encryption.Tools;

namespace Tide.VendorSdk.Classes {
    public class Cipher {
        /*cipher + signature + tag*/
        private static int TagSize => 8;
        private static int EncryptionSize => 32 * 3;
        private static int AsymmetricSize => EncryptionSize * 2 + TagSize;

        public static byte[] Encrypt(string buffer, ulong tag, Tide.Encryption.Ed.Ed25519Key key)
            => Encrypt(Encoding.UTF8.GetBytes(buffer), tag, key);

        public static byte[] Encrypt(byte[] buffer, ulong tag, Tide.Encryption.Ed.Ed25519Key key)
        {
            var toAsymmetricEncrypt =  Pad32(buffer);
            var bufferSymmetric = new byte[0];

            if (buffer.Length > 32)
            {
                var secret = new AesSherableKey();
                bufferSymmetric = secret.Encrypt(buffer);
                toAsymmetricEncrypt = secret.ToByteArray();
            }
            
            var bufferAsymmetric = key.Encrypt(toAsymmetricEncrypt).ToByteArray();
            var tagBuffer = BitConverter.GetBytes(tag);
            var signature = key.EdDSASign(bufferAsymmetric.Concat(tagBuffer).ToArray()).PadLeft(32 * 3);

            var size =
              bufferAsymmetric.Length +
              tagBuffer.Length +
              signature.Length +
              bufferSymmetric.Length;
            var dimension = DimensionBuffer(size);

            var all = new byte[1 + dimension.Length + size];

            all[0] = 1; // version #

            dimension.CopyTo(all, 1);
            var step = dimension.Length + 1;

            bufferAsymmetric.CopyTo(all, step);
            step += bufferAsymmetric.Length;

            tagBuffer.CopyTo(all, step);
            step += tagBuffer.Length;

            signature.CopyTo(all, step);
            step += signature.Length;

            bufferSymmetric.CopyTo(all, step);
            return all;
        }

        public static byte[] Decrypt(byte[] data, Tide.Encryption.Ed.Ed25519Key key)
        {
            var size = data[1] & 127;
            var sizeLength = 0;
            var hasFieldSize = (data[1] & 128) != 0;
            if (hasFieldSize)
            {
                sizeLength = size;
                size = (int) new BigInteger(data.Skip(2).Take(sizeLength).ToArray(), true, true);
            }
            var step = 2 + sizeLength;

            var asymmetricCipher = data.Skip(step).Take(32 * 3).ToArray();
            
            var asymmetricPlain = key.Decrypt(asymmetricCipher);
            step += AsymmetricSize;

            if (size == 200)
                return UnPad32(asymmetricPlain);

            var symmetricKey = AesSherableKey.Parse(asymmetricPlain);
            return symmetricKey.Decrypt(data.Skip(step).ToArray());
        }

        public static byte[] Asymmetric(byte[] data)
        {
            var step = HeadEnd(data);
            return data.Skip(step).Take(AsymmetricSize).ToArray();
        }

        public static byte[] Symmetric(byte[] data)
        {
            var step = HeadEnd(data);
            step += AsymmetricSize;
            
            return data.Skip(step).ToArray();
        }

        public static C25519Cipher CipherFromAsymmetric(byte [] data)
        {
            return C25519Cipher.Parse(data.Take(32 * 3).ToArray());
        }

        public static bool CheckAsymmetric(byte[] data, Tide.Encryption.Ed.Ed25519Key key)
        {
            if (data == null || data.Length != AsymmetricSize)
                return false;

            var cipher = data.Take(EncryptionSize + TagSize).ToArray();
            var signature = data.Skip(EncryptionSize + TagSize).ToArray();

            return key.EdDSAVerify(cipher, signature);
        }
 
        private static int HeadEnd(byte[] data)
        {
            var sizeLength = (data[1] & 128) != 0 ? data[1] & 127 : 0;
            return 2 + sizeLength;
        }

        private static byte[] DimensionBuffer(int size)
        {
            var buffer = (new BigInteger(size)).ToByteArray(true, true);
            return buffer.Length == 1 && buffer[0] < 128 ? buffer
              : (new[] { (byte) ((1 << 7) | buffer.Length) }).Concat(buffer).ToArray();
        }

        public static byte[] Pad32(byte[] data)
        {
            if (data.Length >= 32)
                return data;

            var buffer = new byte[32];
            data.CopyTo(buffer, 32 - data.Length);
            return buffer;
        }

        public static byte[] UnPad32(byte[] data)
        {
            var skip = 0;
            for (var i = 0; i < data.Length; i++)
            {
                if (data[i] != 0)
                    break;
                skip++;
            }

            return skip > 0 ? data.Skip(skip).ToArray() : data;
        }

        public static ulong GetTag(byte[] data)
        {
            var tag = data.Skip(EncryptionSize).Take(TagSize).ToArray();
            return BitConverter.ToUInt64(tag, 0);
        }

        public static C25519Point GetCipherC1(byte[] dataBuffer)
        {
            return C25519Cipher.Parse(dataBuffer.Take(EncryptionSize).ToArray()).C1;
        }
    }
}